app/src/org/commcare/activities/SyncCapableCommCareActivity.java

```diff
@@ -159,6 +166,165 @@ public static void handleSyncUpdate(CommCareActivity activity,
+
+    protected void displayToast(String message) {
+        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
+    }
+
+    @Override
+    public void startBlockingForTask(int id) {
+        super.startBlockingForTask(id);
+        if (isProcessAndSendTaskId(id)) {
+            triggerSyncIconRefresh(TRIGGER_START_SEND_FORMS);
+        } else if (id == DataPullTask.DATA_PULL_TASK_ID) {
+            triggerSyncIconRefresh(TRIGGER_START_DATA_PULL);
+        }
+    }
+
+    @Override
+    public void stopBlockingForTask(int id) {
+        super.stopBlockingForTask(id);
+        if (isProcessAndSendTaskId(id)) {
+            triggerSyncIconRefresh(TRIGGER_END_SEND_FORMS);
+        } else if (id == DataPullTask.DATA_PULL_TASK_ID) {
+            triggerSyncIconRefresh(TRIGGER_END_DATA_PULL);
+        }
+    }
+
+    private static boolean isProcessAndSendTaskId(int id) {
+        return id == ProcessAndSendTask.SEND_PHASE_ID_NO_DIALOG ||
+                id == ProcessAndSendTask.PROCESSING_PHASE_ID_NO_DIALOG ||
+                id == ProcessAndSendTask.PROCESSING_PHASE_ID ||
+                id == ProcessAndSendTask.SEND_PHASE_ID;
+    }
+
+    private void triggerSyncIconRefresh(int trigger) {
+        if (shouldShowSyncItemInActionBar()) {
+            computeSyncState(trigger);
+            rebuildOptionsMenu();
+        }
+    }
+
+    private void computeSyncState(int trigger) {
+        switch(trigger) {
+            case TRIGGER_END_DATA_PULL:
+            case TRIGGER_END_SEND_FORMS:
+            case TRIGGER_NONE:
+                if (SyncDetailCalculations.getNumUnsentForms() > 0) {
+                    syncStateForIcon = SyncState.FORMS_PENDING;
+                } else {
+                    syncStateForIcon = SyncState.UP_TO_DATE;
+                }
+                break;
+            case TRIGGER_START_DATA_PULL:
+                syncStateForIcon = SyncState.PULLING_DATA;
+                break;
+            case TRIGGER_START_SEND_FORMS:
+                syncStateForIcon = SyncState.SENDING_FORMS;
+                break;
+        }
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        if (item.getItemId() == MENU_SYNC) {
+            sendFormsOrSync(true);
+            return true;
+        }
+        return super.onOptionsItemSelected(item);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu);
+        addSyncItemToActionBar(menu);
+        return true;
+    }
+
+    private void addSyncItemToActionBar(Menu menu) {
+        if (shouldShowSyncItemInActionBar() &&
+                Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
+            currentSyncMenuItem = menu.add(MENU_GROUP_SYNC_ACTION, MENU_SYNC, MENU_SYNC, "Sync");
+            currentSyncMenuItem.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
+            switch (syncStateForIcon) {
```

Comment:
Did you consider removing this as a class variable and running computeSyncState() just in time when the value is needed?